dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/handling/DataHandler.java

```diff
@@ -0,0 +1,1161 @@
+        final int valueIndex = grid.getWidth() - 1;
+
+        final List<DimensionalItemObject> basePeriods = params.getPeriods();
+
+        for ( List<Object> row : grid.getRows() )
+        {
+            final List<DimensionalItemObject> dimensionalItems = findDimensionalItems( (String) row.get( dataIndex ),
+                items );
+
+            // Check if the current row's Period belongs to the list of periods from the
+            // original Analytics request.
+            // The row may not have a Period if Period is used as filter.
+            if ( hasPeriod( row, periodIndex ) && isPeriodInPeriods( (String) row.get( periodIndex ), basePeriods ) )
+            {
+                if ( dimensionalItems.size() == 1 )
+                {
+                    addItemBasedOnPeriodOffset( grid, result, periodIndex, valueIndex, row, dimensionalItems );
+                }
+            }
+            else
+            {
+                if ( dimensionalItems.size() == 1 )
+                {
+                    result.put( dimensionalItems.get( 0 ).getDimensionItem(),
+                        new DimensionItemObjectValue( dimensionalItems.get( 0 ), (Double) row.get( valueIndex ) ) );
+                }
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Add the given Indicator values to the Grid.
+     * 
+     * @param params
+     * @param grid
+     * @param dataSourceParams
+     * @param indicator
+     * @param dimensionItems
+     * @param value
+     */
+    private void addIndicatorValuesToGrid( DataQueryParams params, Grid grid, DataQueryParams dataSourceParams,
+        Indicator indicator, List<DimensionItem> dimensionItems, IndicatorValue value )
+    {
+        if ( value != null && satisfiesMeasureCriteria( params, value, indicator ) )
+        {
+            List<DimensionItem> row = new ArrayList<>( dimensionItems );
+
+            row.add( DX_INDEX, new DimensionItem( DATA_X_DIM_ID, indicator ) );
+
+            grid.addRow()
+                .addValues( getItemIdentifiers( row ) )
+                .addValue( getRoundedValue( dataSourceParams, indicator.getDecimals(),
+                    value.getValue() ) );
+
+            if ( params.isIncludeNumDen() )
+            {
+                grid.addValue( getRoundedValue( dataSourceParams, indicator.getDecimals(),
+                    value.getNumeratorValue() ) )
+                    .addValue( getRoundedValue( dataSourceParams, indicator.getDecimals(),
+                        value.getDenominatorValue() ) )
+                    .addValue( getRoundedValue( dataSourceParams, indicator.getDecimals(),
+                        value.getFactor() ) )
+                    .addValue( value.getMultiplier() )
+                    .addValue( value.getDivisor() );
+            }
+        }
+    }
+
+    /**
+     * Calculate the dimensional item offset and adds to the give result map.
+     * 
+     * @param grid
```

Comment:
Add missing javadoc for the method arguments or remove `@param`

Reply:
Done